# Practicumopdrachten Week 2.3 (eerste kans)

## Dependencies importeren

Kopieer in het codeblok hieronder van het vorige practicum de import-code voor de dependencies die het vaakst worden gebruikt om data in te lezen. Geef er ook de gebruikelijke aliassen aan.<br>
Zet eventuele warnings uit.

In [60]:
import pandas as pd
import sqlite3
import warnings
warnings.simplefilter('ignore')

Zet het bestand go_sales.sqlite in hetzelfde mapje als dit notebook.

## Databasetabellen inlezen

Kopieer in het codeblok hieronder van het vorige practicum de code om een connectie met het bestand go_sales.sqlite te maken.

In [61]:
sales_connection = sqlite3.connect('go_sales.sqlite')

Lees van de ingelezen go_sales-database te volgende tabellen in met behulp van "SELECT * FROM *tabel*".
- product
- product_type
- product_line
- sales_staff
- sales_branch
- retailer_site
- country
- order_header
- order_details
- target
- returned_item
- return_reason

In [62]:
tables_to_read = ['product', 'product_type', 'product_line', 'sales_staff', 'sales_branch', 
                  'retailer_site', 'country', 'order_header', 'order_details',
                  'returned_item', 'return_reason']

dataframes = {}

for table in tables_to_read:
    query = f"SELECT * FROM {table}"
    dataframes[table] = pd.read_sql_query(query, sales_connection)

product_df = dataframes['product']
product_type_df = dataframes['product_type']
product_line_df = dataframes['product_line']
sales_staff_df = dataframes['sales_staff']
sales_branch_df = dataframes['sales_branch']
retailer_site_df = dataframes['retailer_site']
country_df = dataframes['country']
order_header_df = dataframes['order_header']
order_details_df = dataframes['order_details']
returned_item_df = dataframes['returned_item']
return_reason_df = dataframes['return_reason']

product_df

PRODUCT_NUMBER INTRODUCTION_DATE PRODUCT_TYPE_CODE PRODUCTION_COST MARGIN  \
0                1         15-2-2011                 1               4    .33   
1               10         15-2-2011                 1              10     .4   
2              100         15-2-2011                17               3     .5   
3              101        15-12-2019                18          305.54    .43   
4              102        10-12-2019                18          380.95    .51   
..             ...               ...               ...             ...    ...   
110             95         15-2-2011                16               3     .5   
111             96         15-2-2011                17           16.43    .28   
112             97          5-3-2013                17              25    .28   
113             98         15-2-2011                17               3     .5   
114             99         15-2-2011                17               2     .6   

       PRODUCT_IMAGE LANGUAGE              PRODUCT_NAME  \
0      P01CE1CG1.jpg       EN       TrailChef Water Bag   
1      P10CE1CG1.jpg       EN        TrailChef Utensils   
2    P100OP4FA17.jpg       EN        Insect Bite Relief   
3    P101GE5IR18.jpg       EN     Hailstorm Steel Irons   
4    P102GE5IR18.jpg       EN  Hailstorm Titanium Irons   
..               ...      ...                       ...   
110   P91OP4SS16.jpg       EN                Sun Shield   
111   P96OP4FA17.jpg       EN        Compact Relief Kit   
112   P96OP4FA17.jpg       EN  Deluxe Family Relief Kit   
113   P98OP4FA17.jpg       EN           Calamine Relief   
114   P99OP4FA17.jpg       EN               Aloe Relief   

                                           DESCRIPTION TRIAL888  
0    Lightweight, collapsible bag to carry liquids ...        T  
1    Spoon, fork and knife set made of a light yet ...        T  
2    The Insect Bite Relief helps the itching and s...        T  
3    Iron is 17-4 stainless steel.  Shafts are grap...        T  
4    Made entirely of pure titanium. The ultimate i...        T  
..                                                 ...      ...  
110  PABA free sunscreen, SPF 30, poison oak and iv...        T  
111  A personal first aid kit is recommended for ev...        T  
112  A complete medical kit suitable for families w...        T  
113  Use the Calamine Relief for allergic skin reac...        T  
114  Perfect for minor burns and sunburn, the aloe ...        T  

[115 rows x 10 columns]

Let op! Voor alle onderstaande opdrachten mag je <u>alleen Python</u> gebruiken, <u>geen SQL!</u>

## Selecties op één tabel zonder functies

Geef een overzicht met daarin de producten en hun productiekosten waarvan de productiekosten lager dan 100 dollar en hoger dan 50 dollar ligt. (2 kolommen, 23 rijen)

In [63]:
product_df['PRODUCTION_COST'] = pd.to_numeric(product_df['PRODUCTION_COST'], errors='coerce')
producten_filtered = product_df.loc[(product_df['PRODUCTION_COST'] > 50) & (product_df['PRODUCTION_COST'] < 100), ['PRODUCT_NAME', 'PRODUCTION_COST']]
producten_filtered

PRODUCT_NAME  PRODUCTION_COST
13                Blue Steel Putter            60.00
14            Blue Steel Max Putter            81.80
17              Course Pro Golf Bag            80.00
24                  Hibernator Lite            60.00
25                       Hibernator            86.00
28  Hibernator Self - Inflating Mat            78.55
31              Hibernator Camp Cot            65.33
32     Canyon Mule Climber Backpack            62.50
56          Granite Climbing Helmet            52.86
58            Husky Harness Extreme            55.00
69                   Granite Hammer            57.13
71        TrailChef Deluxe Cook Set            85.11
81                        Polar Ice            73.33
83                     Polar Sports            80.00
84                       Polar Wave            60.00
85                    Polar Extreme            72.50
88                     Edge Extreme            80.00
91                        Seeker 35            79.19
92                        Seeker 50            92.58
93           TrailChef Double Flame            75.00
94                   Seeker Extreme            94.12
97                   Glacier Deluxe            56.25
98                      Glacier GPS            78.55

Geef een overzicht met daarin de producten en hun marge waarvan de marge lager dan 20 % of hoger dan 60 % ligt. (2 kolommen, 7 rijen) 

In [64]:
product_df['MARGIN'] = pd.to_numeric(product_df['MARGIN'], errors='coerce')
result = product_df[['PRODUCT_NAME', 'MARGIN']][(product_df['MARGIN'] < 0.20) | (product_df['MARGIN'] > 0.60)]
result

PRODUCT_NAME  MARGIN
29                 Hibernator Pad    0.17
32   Canyon Mule Climber Backpack    0.17
100             BugShield Natural    0.67
101               BugShield Spray    0.67
102         BugShield Lotion Lite    0.70
103              BugShield Lotion    0.63
105             BugShield Extreme    0.63

Geef een overzicht met daarin de landen waar met francs wordt betaald. Sorteer de uitkomst op land.  (1 kolom, 3 rijen)

In [65]:
countries_francs = country_df[['COUNTRY', 'CURRENCY_NAME']][country_df['CURRENCY_NAME'] == 'francs']
countries_francs

COUNTRY CURRENCY_NAME
0        France        francs
7       Belgium        francs
19  Switzerland        francs

Geef een overzicht met daarin de verschillende introductiedatums waarop producten met meer dan 50% marge worden geïntroduceerd (1 kolom, 7 rijen) 

In [66]:
product_df['MARGIN'] = pd.to_numeric(product_df['MARGIN'], errors='coerce')
products_high_margin = product_df[['INTRODUCTION_DATE', 'MARGIN']][product_df['MARGIN'] > 0.50]
products_high_margin_unique_dates = products_high_margin.drop_duplicates(subset='INTRODUCTION_DATE')
products_high_margin_unique_dates

INTRODUCTION_DATE  MARGIN
4          10-12-2019    0.51
6          18-12-2019    0.51
7          27-12-2019    0.55
9           13-1-2020    0.55
14         15-12-2019    0.55
79         26-10-2016    0.51
100         15-2-2011    0.67

Geef een overzicht met daarin het eerste adres en de stad van verkoopafdelingen waarvan zowel het tweede adres als de regio bekend is (2 kolommen, 28 rijen)

In [67]:
verkoopafdelingen = sales_branch_df[['ADDRESS1', 'CITY']][(sales_branch_df['ADDRESS2'].notna()) & (sales_branch_df['REGION'].notna())]
verkoopafdelingen

ADDRESS1              CITY
10  Prol. Paseo de la Reforma No. 51  Distrito Federal
11               202-2-3 Hyakunincho             Tokyo
12                     543-225 Asahi        Osaka City
13                  2315 Queen's Ave         Melbourne
15             Avenida Paulista, 333         São Paulo
21                    3 Albany Court        Birmingham
27               Singelgravenplein 4         Amsterdam

Geef een overzicht met daarin de landen waar met dollars (dollars of new dollar) wordt betaald. Sorteer de uitkomst op land. (1 kolom, 4 rijen) 

In [68]:
dollarNewDollar = country_df[['COUNTRY']][(country_df['CURRENCY_NAME'] == 'dollars') | (country_df['CURRENCY_NAME'] == 'new dollar')]
dollarNewDollar

COUNTRY
3          Taiwan
6       Australia
14  United States
15         Canada

Geef een overzicht met daarin beide adressen en de stad van vestigingen van klanten waarvan de postcode begint met een ‘D’ (van duitsland). Filter op vestigingen die een tweede adres hebben. (3 kolommen, 2 rijen) 

In [69]:
clients = retailer_site_df[['ADDRESS1', 'ADDRESS2', 'CITY']][(retailer_site_df['POSTAL_ZONE'].str.startswith('D')) & (retailer_site_df['ADDRESS2'].notnull())]
clients

ADDRESS1       ADDRESS2       CITY
205  Röntgenstraße 90  3. Tür rechts  Frankfurt
351   Grubesallee 141       4. Stock    Hamburg

## Selecties op één tabel met functies

Geef het totaal aantal producten dat is teruggebracht (1 waarde) 

In [70]:
returned_item_df['RETURN_QUANTITY'] = pd.to_numeric(returned_item_df['RETURN_QUANTITY'])
total_returned_items = [returned_item_df['RETURN_QUANTITY'].sum()]
total_returned_items

[14306]

Geef het aantal regio’s waarin verkoopafdelingen gevestigd zijn. (1 waarde)

In [71]:
region_amount = sales_branch_df['REGION'].count()
region_amount

15

Maak 3 variabelen:
- Een met de laagste
- Een met de hoogste
- Een met de gemiddelde (afgerond op 2 decimalen)

marge van producten (3 kolommen, 1 rij) 

In [72]:
laagste = product_df['MARGIN'].min()
hoogste = product_df['MARGIN'].max()
gemiddelde = product_df['MARGIN'].mean()
resultaten = pd.DataFrame({
    'Laagste': [laagste],
    'Hoogste': [hoogste],
    'gemiddelde': [gemiddelde]
})
resultaten

Laagste  Hoogste  gemiddelde
0     0.17      0.7    0.396609

Geef het aantal vestigingen van klanten waarvan het 2e adres niet bekend is (1 waarde)

In [73]:
nonAvailableAdress = retailer_site_df['ADDRESS2'].isnull().sum()
nonAvailableAdress

286

Geef de gemiddelde kostprijs van de verkochte producten waarop korting (unit_sale_price < unit_price) is verleend (1 waarde) 

In [74]:
order_details_df['UNIT_COST'] = pd.to_numeric(order_details_df['UNIT_COST'])
averageCostPriceDiscount = order_details_df[['UNIT_COST']][(order_details_df['UNIT_SALE_PRICE'] < order_details_df['UNIT_PRICE'])].mean()
averageCostPriceDiscount

UNIT_COST    121.994812
dtype: float64

Geef een overzicht met daarin het aantal medewerkers per medewerkersfunctie (2 kolommen, 7 rijen) 

In [75]:
medewerkersPerFunctie = sales_staff_df.groupby('POSITION_EN')['SALES_STAFF_CODE'].count()
medewerkersPerFunctie

POSITION_EN
Branch Manager                  19
District Manager                 4
General Manager                  1
Level 1 Sales Representative    13
Level 2 Sales Representative    39
Level 3 Sales Representative    24
Regional Manager                 2
Name: SALES_STAFF_CODE, dtype: int64

Geef een overzicht met daarin per telefoonnummer het aantal medewerkers dat op dat telefoonnummer bereikbaar is. Toon alleen de telefoonnummer waarop meer dan 4 medewerkers bereikbaar zijn. (2 kolommen, 10 rijen) 

In [76]:
medewerkers_per_telefoonnummer = sales_staff_df.groupby('WORK_PHONE')['SALES_STAFF_CODE'].count().reset_index()
telefoonnummers_meer_dan_4_medewerkers = medewerkers_per_telefoonnummer[medewerkers_per_telefoonnummer['SALES_STAFF_CODE'] > 4]
telefoonnummers_meer_dan_4_medewerkers

WORK_PHONE  SALES_STAFF_CODE
4     +(44) 121 3505267                 5
10  +(61) 03 2982  4242                 5
11  +31 (0)20 692 93 94                 5
13    +33 1 68 94 52 20                 5
18     1 (206) 292-0012                 5
19     1 (305) 557-4810                 6
20     1 (310) 281-5722                 5
21     1 (403) 232-5986                 5
22     1 (416) 493-5595                 5
23     1 (617) 268-6754                 6

## Selecties op meerdere tabellen zonder functies

Geef een overzicht met daarin het eerste adres en de stad van vestigingen van klanten uit ‘Netherlands’ (2 kolommen, 20 rijen) 

In [77]:
klantenUitNederland = retailer_site_df[['ADDRESS1', 'CITY']][retailer_site_df['COUNTRY_CODE'] == '7']
klantenUitNederland

ADDRESS1          CITY
164                   Beets 36         Beets
165               Perestraat 1      Deventer
166  Laan van Meerdervoort 966      Den Haag
167             Kalfjeslaan 47    Amstelveen
168               Startbaan 11    Amstelveen
169      Noorderspoorsingel 28     Groningen
170              Hogehilweg 13     Amsterdam
171             Spoorstraat 43    Varsseveld
173             Westhavenweg 8     Amsterdam
174              Westzeedijk 4        Muiden
175            Gooiergracht 91         Laren
176            Hoofdstraat 150   Pieterburen
177                Olde Diek 2         Laren
178     Amsterdamsestraatweg 8      Maarssen
179        Krayenhoffstraat 11     Amsterdam
180         Overschiestraat 19     Amsterdam
181              Havenbuurt 12        Marken
182               Duivendaal 1    Wageningen
184              Voorstraat 37  Nieuweschans
185         Stadhouderslaan 18     Rotterdam

Geef een overzicht met daarin de productnamen die tot het producttype ‘Eyewear’ behoren. (1 kolom, 5 rijen) 

In [78]:
eyewareProducts = product_df['PRODUCT_NAME'][product_df['PRODUCT_TYPE_CODE'] == '11'] 
eyewareProducts

80        Polar Sun
81        Polar Ice
83     Polar Sports
84       Polar Wave
85    Polar Extreme
Name: PRODUCT_NAME, dtype: object

Geef een overzicht met daarin alle unieke eerste adressen van klantvestigingen en de voornaam en achternaam van de verkopers die ‘Branch Manager’ zijn en aan deze vestigingen hebben verkocht (3 kolommen, 1 rij) 

In [79]:
merged = pd.merge(sales_staff_df, sales_branch_df, on='SALES_BRANCH_CODE', how='left')
branch_managers = merged[merged['POSITION_EN'] == 'Branch Manager']
unique_addresses = merged['ADDRESS1'].unique()
result_df = branch_managers[branch_managers['ADDRESS1'].isin(unique_addresses)][['ADDRESS1', 'FIRST_NAME', 'LAST_NAME']]
result_df

ADDRESS1 FIRST_NAME      LAST_NAME
7                     Victoria House     Walter         Taylor
11                    3 Albany Court      Polly         Linton
19                 Interleuvenlaan 2      Gracy        Gellens
27                    Schwabentor 35      Frank          Fuchs
30                  Leopoldstraße 36      Fritz         Hirsch
33                   Piazza Duomo, 1      Maria      Iacobucci
38               Singelgravenplein 4       Kick        Kalkman
43               Isafjordsgatan 30 C      Karin      Bergström
44     75, rue du Faubourg St-Honoré      Denis           Pagé
47      7800, 756 - 6th Avenue. S.W.      Sally          White
50                  789 Yonge Street    Georges  Saint-Germain
56            1288 Dorchester Avenue       Alex      Rodriguez
60                   299 Yale Avenue   Samantha          Floyd
65                    10032 NW 186th      Julie          Olsen
69        1288 South Barrington Ave.      Corey         Wright
80               6c, rue de l'Église      Julie        Béchard
84  Prol. Paseo de la Reforma No. 51      Lucía          Reyna
91                  2315 Queen's Ave      Alice         Walter
98             Avenida Paulista, 333     Bayard          Lopes

Geef een overzicht met daarin van de verkopers hun functie en indien zij iets hebben verkocht de datum waarop de verkoop heeft plaatsgevonden. Laat alleen de verschillende namen van de posities zien van de verkopers die het woord ‘Manager’ in hun positienaam hebben staan. (2 kolommen, 9 rijen) 

In [80]:
merged_df = pd.merge(sales_staff_df, sales_branch_df, on='SALES_BRANCH_CODE', how='left')
merged_order_df = pd.merge(merged_df, order_header_df, on=['SALES_STAFF_CODE', 'SALES_STAFF_CODE'], how='inner')
manager_positions = merged_order_df[merged_order_df['POSITION_EN'].str.contains('Manager')]
result_df = manager_positions[['POSITION_EN', 'ORDER_DATE']]
result_df

POSITION_EN  ORDER_DATE
5300  Branch Manager  2022-09-09
5301  Branch Manager  2022-10-31
5302  Branch Manager  2022-12-13
5303  Branch Manager  2022-03-13
5304  Branch Manager  2022-01-09

Geef een overzicht met daarin de verschillende namen van producten en bijbehorende namen van producttypen van de producten waarvoor ooit meer dan 750 stuks tegelijk verkocht zijn. (2 kolommen, 9 rijen) 

In [81]:
order_details_df['QUANTITY'] = pd.to_numeric(order_details_df['QUANTITY'], errors='coerce')
high_quantity_orders = order_details_df[order_details_df['QUANTITY'] > 750]
merged_product_info = pd.merge(high_quantity_orders, product_df, on='PRODUCT_NUMBER', how='inner')
result_df = merged_product_info[['PRODUCT_NAME', 'PRODUCT_TYPE_CODE']]
result_df = result_df.drop_duplicates()
result_df

PRODUCT_NAME PRODUCT_TYPE_CODE
0          Sun Shelter 30                16
1         BugShield Spray                15
3   BugShield Lotion Lite                15
5               Firefly 2                 5
7        BugShield Lotion                15
9       BugShield Extreme                15
10         Sun Shelter 15                16
11             Sun Shield                16
15               Star Peg                 2

Geef een overzicht met daarin de productnamen waarvan ooit meer dan 40% korting is verleend. De formule voor korting is: (unit_price - unit_sale_price) / unit_price (1 kolom, 8 rijen) 

In [82]:
order_details_df['UNIT_PRICE'] = pd.to_numeric(order_details_df['UNIT_PRICE'], errors='coerce')
order_details_df['UNIT_SALE_PRICE'] = pd.to_numeric(order_details_df['UNIT_SALE_PRICE'], errors='coerce')
merged_df = pd.merge(product_df, order_details_df, on='PRODUCT_NUMBER', how='inner')
merged_df['DISCOUNT_PERCENTAGE'] = ((merged_df['UNIT_PRICE'] - merged_df['UNIT_SALE_PRICE']) / merged_df['UNIT_PRICE']) * 100
high_discount_products = merged_df[merged_df['DISCOUNT_PERCENTAGE'] > 40]
result_df = high_discount_products[['PRODUCT_NAME']].drop_duplicates()
result_df

PRODUCT_NAME
2765   Hailstorm Titanium Woods Set
37446             BugShield Natural
37967               BugShield Spray
38306         BugShield Lotion Lite
38772              BugShield Lotion
39561             BugShield Extreme
40799                Sun Shelter 15
41194                Sun Shelter 30

Geef een overzicht met daarin de retourreden van producten waarvan ooit meer dan 90% van de aangeschafte hoeveelheid is teruggebracht (return_quantity/quantity). (1 kolom, 3 rijen) 

In [83]:
merged_returns_reason_df = pd.merge(returned_item_df, return_reason_df, on='RETURN_REASON_CODE', how='left')
merged_returns_df = pd.merge(order_details_df, merged_returns_reason_df, on='ORDER_DETAIL_CODE', how='inner')
merged_returns_df['QUANTITY'] = pd.to_numeric(merged_returns_df['QUANTITY'], errors='coerce')
merged_returns_df['RETURN_QUANTITY'] = pd.to_numeric(merged_returns_df['RETURN_QUANTITY'], errors='coerce')
merged_returns_df['RETURN_PERCENTAGE'] = (merged_returns_df['RETURN_QUANTITY'] / merged_returns_df['QUANTITY']) * 100
high_return_products = merged_returns_df[merged_returns_df['RETURN_PERCENTAGE'] > 90]
result_df = high_return_products[['RETURN_DESCRIPTION_EN']].drop_duplicates()
result_df

RETURN_DESCRIPTION_EN
0    Wrong product shipped
1   Unsatisfactory product
15   Wrong product ordered

## Selecties op meerdere tabellen met functies

Geef een overzicht met daarin per producttype het aantal producten die tot dat producttype behoren. (2 kolommen, 21 rijen) 

In [84]:
merged_products_orders_df = pd.merge(product_df, order_details_df, on='PRODUCT_NUMBER', how='left')
product_type_counts = merged_products_orders_df.groupby('PRODUCT_TYPE_CODE')['PRODUCT_NUMBER'].nunique().reset_index()
product_type_counts.columns = ['PRODUCT_TYPE_CODE', 'NUMBER_OF_PRODUCTS']
product_type_counts

PRODUCT_TYPE_CODE  NUMBER_OF_PRODUCTS
0                  1                  10
1                 10                   5
2                 11                   5
3                 12                   5
4                 13                   4
5                 14                   4
6                 15                   5
7                 16                   5
8                 17                   5
9                 18                   4
10                19                   4
11                 2                   6
12                20                   3
13                21                   4
14                 3                   7
15                 4                   6
16                 5                  12
17                 6                   4
18                 7                   4
19                 8                   7
20                 9                   6

Geef een overzicht met daarin per land het aantal vestigingen van klanten die zich in dat land bevinden. (2 kolommen, 21 rijen) 

In [85]:
merged_retailer_sites_df = pd.merge(retailer_site_df, country_df, left_on='COUNTRY_CODE', right_on='COUNTRY_CODE', how='inner')
retailer_sites_by_country = merged_retailer_sites_df.groupby(['COUNTRY'])['RETAILER_SITE_CODE'].nunique().reset_index()
retailer_sites_by_country

COUNTRY  RETAILER_SITE_CODE
0        Australia                  10
1          Austria                  10
2          Belgium                  10
3           Brazil                   1
4           Canada                  31
5            China                  10
6          Denmark                   6
7          Finland                   5
8           France                  29
9          Germany                  32
10           Italy                  12
11           Japan                  25
12           Korea                  10
13          Mexico                  10
14     Netherlands                  20
15           Spain                   9
16          Sweden                  12
17     Switzerland                  13
18          Taiwan                  10
19  United Kingdom                  25
20   United States                 101

Geef een overzicht met daarin van de producten behorend tot het producttype ‘Cooking Gear’ per productnaam de totaal verkochte hoeveelheid en de gemiddelde verkoopprijs. Sorteer de uitkomst op totaal verkochte hoeveelheid. (4 kolommen, 10 rijen) 

In [86]:
cooking_gear_products = product_df[product_df['PRODUCT_TYPE_CODE'] == '1']
merged_df = pd.merge(cooking_gear_products, order_details_df, on='PRODUCT_NUMBER', how='inner')
merged_df = pd.merge(merged_df, product_type_df, on='PRODUCT_TYPE_CODE', how='inner')
result_df = merged_df.groupby('PRODUCT_NAME').agg({
    'QUANTITY': 'sum',
    'UNIT_SALE_PRICE': 'mean',
    'PRODUCT_TYPE_EN': 'first'
}).reset_index()
result_df = result_df.sort_values(by='QUANTITY', ascending=False)
result_df

PRODUCT_NAME  QUANTITY  UNIT_SALE_PRICE PRODUCT_TYPE_EN
9        TrailChef Water Bag     41770         5.999266    Cooking Gear
2              TrailChef Cup     31834         6.945976    Cooking Gear
0          TrailChef Canteen     31712        11.930119    Cooking Gear
6      TrailChef Kitchen Kit     25964        22.210198    Cooking Gear
8         TrailChef Utensils     15928        17.132397    Cooking Gear
4     TrailChef Double Flame     14510       131.348013    Cooking Gear
1         TrailChef Cook Set     12700        51.204554    Cooking Gear
7     TrailChef Single Flame     10614        63.572397    Cooking Gear
5           TrailChef Kettle      9900        12.130783    Cooking Gear
3  TrailChef Deluxe Cook Set      3744       120.303667    Cooking Gear

Geef een overzicht met daarin van de verkopers behorend tot de verkoopafdeling uit ‘Amsterdam’ de voornaam en achternaam van de verkoper en zijn laagste en hoogste verkooptarget van 2006. (4 kolommen, 4 rijen) 

In [87]:
merged_df = pd.merge(sales_staff_df, sales_branch_df, on='SALES_BRANCH_CODE', how='left')
sales_staff_amsterdam_df = merged_df[merged_df['CITY'] == 'Amsterdam']
result_df = sales_staff_amsterdam_df[['FIRST_NAME', 'LAST_NAME', 'EXTENSION']]
result_df

FIRST_NAME         LAST_NAME EXTENSION
35    Belinda  Jansen-Velasquez       187
36      Ellen           Shapiro       179
37        Jan         Haverkamp       163
38       Kick           Kalkman       197
39      Maria          Laponder       144

Geef een overzicht met daarin per land de naam van het land, de naam van de stad waar de verkoopafdeling is gevestigd (noem de kolomnaam in het overzicht ‘verkoper’) en het aantal steden waar zich klanten bevinden in dat land (noem de kolomnaam in het overzicht ‘klanten’) (3 kolommen, 29 rijen) 

In [88]:
merged_sales_branch = pd.merge(sales_branch_df, country_df, on='COUNTRY_CODE', how='left')
grouped_data = merged_sales_branch.groupby(['COUNTRY', 'CITY']).size().reset_index(name='Klanten')
result_df = grouped_data[['COUNTRY', 'CITY', 'Klanten']]
result_df.columns = ['Land', 'Verkoper', 'Klanten']
result_df

Land          Verkoper  Klanten
0        Australia         Melbourne        1
1          Austria              Wien        1
2          Belgium          Heverlee        1
3           Brazil         São Paulo        1
4           Canada           Calgary        1
5           Canada           Toronto        1
6            China          Shanghai        1
7          Finland            Kuopio        1
8           France              Lyon        1
9           France             Paris        1
10         Germany           Hamburg        1
11         Germany           München        1
12           Italy            Milano        1
13           Japan        Osaka City        1
14           Japan             Tokyo        1
15           Korea             Seoul        1
16          Mexico  Distrito Federal        1
17     Netherlands         Amsterdam        1
18           Spain            Bilbao        1
19          Sweden             Kista        1
20     Switzerland            Zürich        1
21          Taiwan         San Chung        1
22  United Kingdom        Birmingham        1
23  United Kingdom            London        1
24   United States            Boston        1
25   United States       Los Angeles        1
26   United States             Miami        1
27   United States           Seattle        1

## Pythonvertalingen van SUBSELECT en UNION met o.a. for-loops

Geef een overzicht met daarin de voornaam en de achternaam van de medewerkers die nog nooit wat hebben verkocht (2 kolommen, 25 rijen) 

In [89]:
# Find all sales staff codes in order_header_df
existing_codes = order_header_df['SALES_STAFF_CODE']

# Filter sales_staff_df to exclude codes in existing_codes
staff_without_sales = sales_staff_df[~sales_staff_df['SALES_STAFF_CODE'].isin(existing_codes)]
selected_columns = ['FIRST_NAME', 'LAST_NAME']
filtered_staff = staff_without_sales[selected_columns]

# Print the DataFrame
filtered_staff

FIRST_NAME      LAST_NAME
7      Walter         Taylor
8    Penelope        Tamrine
11      Polly         Linton
19      Gracy        Gellens
26       Else         Mörike
27      Frank          Fuchs
30      Fritz         Hirsch
31       Jörg          Kunze
33      Maria      Iacobucci
38       Kick        Kalkman
43      Karin      Bergström
44      Denis           Pagé
47      Sally          White
49      Frank        Bretton
50    Georges  Saint-Germain
56       Alex      Rodriguez
60   Samantha          Floyd
65      Julie          Olsen
69      Corey         Wright
76    Estelle        Johnson
77       Greg        Belding
80      Julie        Béchard
84      Lucía          Reyna
89     Kazumi        Uragome
91      Alice         Walter

Geef een overzicht met daarin het aantal producten waarvan de marge lager is dan de gemiddelde marge van alle producten samen. Geef in het overzicht tevens aan wat de gemiddelde marge is van dit aantal producten waarvan de marge lager dan de gemiddelde marge van alle producten samen is. (1 kolom, 2 rijen) 

In [90]:
numericMargin = pd.to_numeric(product_df['MARGIN'])
total_margin = numericMargin.sum()
# shape wordt gebruikt om het aantal rijen en kolommen te bepalen door[0] bepaal je aantal rijen en met [1] aantal kolommen
total_Products = product_df['PRODUCT_NUMBER'].count()
avg_margin_product = total_margin/total_Products
low_margin_products = product_df['MARGIN'] < avg_margin_product

total_low_margin = low_margin_products.sum()
data = {
    "Average Margin": [avg_margin_product],
    "Products Below Avg": [low_margin_products.sum()]
}

overview_df = pd.DataFrame(data)

overview_df

Average Margin  Products Below Avg
0        0.396609                  59

Geef een overzicht met daarin de namen van de producten die voor meer dan 500 (verkoopprijs) zijn verkocht maar nooit zijn teruggebracht. (1 kolom, 13 rijen) 

In [91]:
products_over500 = product_df["PRODUCTION_COST"] * (1 +product_df["MARGIN"]) > 500
over500_products_df = product_df[products_over500]
over500_product_names = over500_products_df["PRODUCT_NAME"]
over500_product_names #klopt niet met de 13 rijen die verwacht worden en ik moet ook nog de check uitvoeren of ze nog nooit zijn teruggebracht

4              Hailstorm Titanium Irons
6         Lady Hailstorm Titanium Irons
7          Hailstorm Titanium Woods Set
8             Hailstorm Steel Woods Set
9     Lady Hailstorm Titanium Woods Set
10       Lady Hailstorm Steel Woods Set
19                            Star Dome
21                         Star Gazer 3
22                         Star Gazer 6
55                       Husky Rope 200
Name: PRODUCT_NAME, dtype: object

Geef een overzicht met daarin per (achternaam van) medewerker of hij/zij manager is of niet, door deze informatie toe te voegen als extra 'Ja/Nee'-kolom.<br>
Hint: gebruik een for-loop waarin je o.a. bepaalt of het woord 'Manager' in de functie (position_en) staat. (2 kolommen, 102 rijen).

In [92]:
naam_functie = sales_staff_df['LAST_NAME'], sales_staff_df['POSITION_EN']
naam_functie

(0      Savolainen
 1             Kim
 2            Choi
 3             Kao
 4            Fang
           ...    
 97       Iglesias
 98          Lopes
 99         Barros
 100     Guimarães
 101      Lindholm
 Name: LAST_NAME, Length: 102, dtype: object,
 0      Level 2 Sales Representative
 1      Level 2 Sales Representative
 2      Level 3 Sales Representative
 3      Level 2 Sales Representative
 4      Level 3 Sales Representative
                    ...             
 97     Level 2 Sales Representative
 98                   Branch Manager
 99     Level 3 Sales Representative
 100    Level 2 Sales Representative
 101    Level 3 Sales Representative
 Name: POSITION_EN, Length: 102, dtype: object)

Met de onderstaande code laat je Python het huidige jaar uitrekenen.

In [93]:
from datetime import date
date.today().year

2024

Met de onderstaande code selecteer je op een bepaald jaartal uit een datum.

In [94]:
from datetime import datetime

date_str = '16-8-2013'
date_format = '%d-%m-%Y'
date_obj = datetime.strptime(date_str, date_format)

date_obj.year

2013

Geef met behulp van bovenstaande hulpcode een overzicht met daarin op basis van het aantal jaar dat iemand in dienst is of een medewerker ‘kort in dienst’ (minder dan 25 jaar in dienst) of een ‘lang in dienst’ (groter gelijk dan 12 jaar in dienst) is. Geef daarbij per medewerker in een aparte kolom zowel ‘kort in dienst’ als ‘lang in dienst’ aan. Gebruik (wederom) een for-loop.<br>
(2 kolommen, 102 rijen) 